File: Solution/Programmers/순위.py
def solution(n, results):
    wins = {}  # 내가 이긴 선수들
    loses = {}  # 내가 진 선수들
    for i in range(1, n+1):
        wins[i] = set()
        loses[i] = set()
    
    for i in range(1, n+1):
        for result in results:
            if result[0] == i:
                wins[i].add(result[1])
            if result[1] == i:
                loses[i].add(result[0])
        for winner in loses[i]:
            wins[winner].update(wins[i])
        for loser in wins[i]:
            loses[loser].update(loses[i])
    
    ans = 0
    for i in range(1, n+1):
        if len(wins[i]) + len(loses[i]) == n-1:
            ans += 1
    return ans

File: Solution/Programmers/순위.js
function solution(n, results) {
  const graph = Array.from(Array(n + 1), () => Array(n + 1).fill(false));
  results.map(result => {
    const [a, b] = result;
    graph[a][b] = 1;
    graph[b][a] = -1;
    graph[a][a] = 0;
    graph[b][b] = 0;
  });

  const range = [...Array(n).keys()].map(e => e + 1);

  for (const mid of range) {
    for (const a of range) {
      for (const b of range) {
        // a가 mid를 이기고, mid가 b를 이기면 a가 b를 이긴 것과 같음
        if (graph[a][mid] === 1 && graph[mid][b] === 1) graph[a][b] = 1;
        if (graph[a][mid] === -1 && graph[mid][b] === -1) graph[a][b] = -1;
      }
    }
  }

  let ans = 0;
  for (const i of range) {
    let flag = true;
    for (const j of range) {
      if (graph[i][j] === false) {
        flag = false;
        break;
      }
    }
    if (flag) ans++;
  }

  return ans;
}
